### PR TITLE
Add each_hunk method

### PR DESCRIPTION
**Pull request by ouroboros8**
This method allows iteration over "change hunks", groups of contiguous
additions or deletions surrounded by unchanged context lines.

Don't have an issue raised for this, hope that's ok!

### CODE REVIEW

File: lib/diffy/diff.rb [hunk 1]
```diff
@@ -114,6 +114,29 @@ def each_chunk
       end
     end
 
+    def each_hunk
+      # Iterate over "change hunks": contigous hunks of context around a
+      # change. Only works with unified diffs:
+      # https://en.wikipedia.org/wiki/Diff_utility#Unified_format
+      if not @options[:diff] =~ /-U/
+        raise "Diffy only supports each_hunk for unified diffs."
+      end
```
Comment (ouroboros8) [inline lib/diffy/diff.rb]: This bit is a bit of a hack I'm afraid, but I think just crashing out with an error is preferable to silently failing when the diff is of the wrong type. Open to suggestions though!

File: lib/diffy/diff.rb [hunk 2]
```diff
@@ -114,6 +115,31 @@ def each_chunk
...
+      hunks = []
+      while i = remaining_diff.index(header_pattern, 1)
+        hunks.push(remaining_diff[0..i-1])
+        remaining_diff = remaining_diff[i..-1]
+      end
+      hunks.push(remaining_diff)
+
+      if not @options[:include_diff_info]
+        # Drop the diff header and each hunk's hunk header
+        hunks = hunks[1..-1].map { | hunk | hunk.sub(/^.*\n/, '') }
```
Comment (ouroboros8) [inline lib/diffy/diff.rb]: oops that substitution is wrong

**Comment by ouroboros8** [comment]
Have found an issue relating to the way that the `:allow_empty_diff` option

**Comment by samg** [comment]
Maybe take a look at `Diffy::Diff#each_chunk` https://github.com/samg/diffy/blob/master/lib/diffy/diff.rb#L97-L108

I think it may do something similar.

**Pull request closed without merging**